### DISCUSSION TITLE
How to use transactons properly?

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4636] [Q&A] [answered]
Hello. While performing a refactoring of my project that uses linq2db and SqlCe I've found that I use transactions in several ways:
```
var fileName = "test.sdf";
var connectionStringFromat =
    "Persist Security Info = False; Data Source = {0}; File Mode = 'Read Write'; Max Database Size = 2048; SSCE:Default Lock Timeout=300000; Flush interval = 1";
using (var connection = new SqlCeConnection(string.Format(connectionStringFromat, fileName)))
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
using (var transaction = connection.BeginTransaction())
{
    try
    {
        //modify some data via db.Insert() / db.Update() / db.BulkCopy(), etc.
       transaction.Commit(CommitMode.Immediate);
    }
    catch (Exception ex)
    {
        Console.WriteLine(ex);
        transaction.Rollback();
    }
}
```
or I can start a transaction before creating a DataConnection:
```
using (var connection = new SqlCeConnection(string.Format(connectionStringFromat, fileName)))
using (var transaction = connection.BeginTransaction())
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
```
or I can start a transaction and pass it into a DataConnection constructor:
```
using (var connection = new SqlCeConnection(string.Format(connectionStringFromat, fileName)))
using (var transaction = connection.BeginTransaction())
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), transaction))
```
or instead of starting a transaction for an underlying connection I can start it via DataConnection class itself:
```
using (var connection = new SqlCeConnection(string.Format(connectionStringFromat, fileName)))
using (var db = new DataConnection(SqlCeTools.GetDataProvider(), connection))
using (var transaction = db.BeginTransaction())
```
What is preferable way to use transactions an why? Are there any recommendations how to use them correctly?
Thank you.

**Comment by MaceWindu** [comment, marked as answer]
I would recommend use methods, that will tell `Linq To DB` about transaction existence (two last examples). While in most cases we don't care about transaction, there are currently two situations where it makes difference:
- some database providers doesn't implement `CommandBehavior.SchemaOnly` and we need to create explicit transaction when load schema for procedures for scaffold (or database schema API use in general). This doesn't apply to SqlCe (currently it is done for Access, SAP HANA and ASE)
- for multi-statement eager load/client-side group-by queries we wrap them into transaction if current connection doesn't have open transaction already

**Reply by alexey-leonovich** [reply]
Thank you.
